Add pipeTo support for readable byte streams

youennf opened the pull request:
#### a1b79c5541d9549f0c200870202cd8a64af7a301
<pre>
Add pipeTo support for readable byte streams
<a href="https://bugs.webkit.org/show_bug.cgi?id=299438#">https://bugs.webkit.org/show_bug.cgi?id=299438#</a>
<a href="https://rdar.apple.com/problem/161244992">rdar://problem/161244992</a>

Reviewed by Chris Dumez.

We implement the pipeTo algorithm in C++ using a default reader.
This is the same algorithm as for regular streams so it might replace the JS built-in implementation in a follow-up for regular streams.

The implementation is done via PipeToState which stores the necessary state values needed for pipeTo.
It notably keeps a reference to the source and destination to propagate the errors backward and forward.
It also keeps reference to the reader and writer used for pipeTo.

The implementation follows closely the specification so the pipeTo operations are all done in the thread where live the streams.

Covered by rebased tests.

Canonical link: <a href="https://commits.webkit.org/301030@main">https://commits.webkit.org/301030@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73be3e274a70a3d81d8d95368b5d4e26fae2332e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124741 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131587 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76660 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2b52369b-6002-4813-8fa8-218919bb545e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126618 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45117 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94909 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62966 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7930097e-954c-40a0-ad05-0ba8de600c42) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111561 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75478 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e67e6bd7-a923-406e-b5fe-4eeb50989139) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34918 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29715 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75063 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105744 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134252 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51589 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39400 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103386 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103157 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26258 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48525 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26803 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48586 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51462 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57260 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50856 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54211 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52550 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0690211139df05d1c29c1ec3aa4ac1904967f570. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/51c19391b4737882af89c59bf535516f93451274. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7d924ee9f3d736dd2873ac287344737b8d9ebff3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0d257c3870f669e882d6194c9cec8dafde25a18e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8bdcc8fdd9a16f4494d3f70d3daa38b91c9a07ae. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a10527cec7d3ee8a748485583aadfeb123ee5484. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8b7420a679b71c08a59cafceefb9f6c1857e6c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/73be3e274a70a3d81d8d95368b5d4e26fae2332e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51565

youennf committed:
Add pipeTo support for readable byte streams
https://bugs.webkit.org/show_bug.cgi?id=299438#
rdar://problem/161244992

Reviewed by Chris Dumez.

We implement the pipeTo algorithm in C++ using a default reader.
This is the same algorithm as for regular streams so it might replace the JS built-in implementation in a follow-up for regular streams.

The implementation is done via PipeToState which stores the necessary state values needed for pipeTo.
It notably keeps a reference to the source and destination to propagate the errors backward and forward.
It also keeps reference to the reader and writer used for pipeTo.

The implementation follows closely the specification so the pipeTo operations are all done in the thread where live the streams.

Covered by rebased tests.

Canonical link: https://commits.webkit.org/301030@main
